### PR TITLE
BSL/VSL validation code cleanup

### LINKED ISSUE #1129: Clean up duplicate logic for validating backupstoragelocations and volumesnapshotlocations in dpa
https://github.com/kaovilai/oadp-operator/blob/27cccabbd2e845da21ef49489e7506f10c30696d/controllers/validator.go#L19

### PR DESCRIPTION
**Pull request by stillalearner**
Fixes #1129

### CODE REVIEW

File: controllers/bsl.go
Comment (kaovilai) [inline controllers/bsl.go:20]: Note to self: This is already in validator.go

File: controllers/vsl.go
Comment (kaovilai) [inline controllers/vsl.go:113]: Note to self: This is already in validator.go

**Comment by kaovilai** [comment]
oh and unit tests failing, can assist with that if needed.

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go]: done

File: controllers/bsl_test.go
Comment (kaovilai) [inline controllers/bsl_test.go:1125]: What's the `testing` key here for?

Comment (stillalearner) [reply controllers/bsl_test.go:1125]: haha, can be removed, took this yaml from validator_test file. let me make it "cloud"

Comment (stillalearner) [reply controllers/bsl_test.go]: fixed.

Comment (kaovilai) [reply controllers/bsl_test.go]: ahh sorry. didn't notice the keys were specified in cloudstorage

Comment (kaovilai) [reply controllers/bsl_test.go]: before was probably better if there's a check for `dummy_data2` being used correctly..

Comment (kaovilai) [reply controllers/bsl_test.go]: but since there's not, up to you.

**Review by kaovilai — APPROVED** [review]

**Review by kaovilai — APPROVED** [review]

**Comment by mateusoliveira43** [comment]
can we hold this one? I think after this ones merges https://github.com/openshift/oadp-operator/pull/1294, this will need rebase (or vice versa)

File: controllers/bsl.go
Comment (mateusoliveira43) [inline controllers/bsl.go:389]: For consistency, I would change message here to **"BackupLocation must have velero or bucket configuration"**

Since this check was in validator.go, it would run prior to this. So, I think this message was never presented to the user

File: controllers/bsl.go
Comment (mateusoliveira43) [inline controllers/bsl.go:23]: can this for be merged with the following?

something like
```go
       // Ensure BSL is a valid configuration
	// First, check for provider and then call functions based on the cloud provider for each backupstoragelocation configured
	numDefaultLocations := 0
	for _, bslSpec := range dpa.Spec.BackupLocations {
           	if err := r.ensureBackupLocationHasVeleroOrCloudStorage(&bsl); err != nil {
			return false, err
		}

		if err := r.ensurePrefixWhenBackupImages(&dpa, &bsl); err != nil {
			return false, err
		}

		if err := r.ensureSecretDataExists(&dpa, &bsl); err != nil {
			return false, err
		}
             if bslSpec.Velero != nil {
                ....
```

File: controllers/bsl.go
Comment (mateusoliveira43) [inline controllers/bsl.go:413]: after rebase on top of https://github.com/openshift/oadp-operator/pull/1294, please change lines 412 and 413 to 
```go
// Check if the Velero feature flag 'no-secret' is not set
if !dpa.Spec.Configuration.Velero.HasFeatureFlag("no-secret") {
```

File: controllers/bsl_test.go
Comment (mateusoliveira43) [inline controllers/bsl_test.go:1714]: would also add the error message as well here; then, in cases where we want error, we would check if the message is the expected one

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go:389]: fixed.

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go:23]: yes, i was thinking of that earlier but kept it separate for better readability. But as we have proper functions now, it looks readable like this as well.

Comment (stillalearner) [reply controllers/bsl.go:23]: fixed

File: controllers/bsl_test.go
Comment (stillalearner) [reply controllers/bsl_test.go]: fixed. Thank you!

File: controllers/bsl.go
Comment (mateusoliveira43) [inline controllers/bsl.go:19]: This one we can remove as well, it is duplication of https://github.com/openshift/oadp-operator/blob/8890679c9b2c6866860badbd3c870ff3392a5ec8/controllers/validator.go#L37-L41

Maybe is even better to bring this part of code https://github.com/openshift/oadp-operator/blob/8890679c9b2c6866860badbd3c870ff3392a5ec8/controllers/validator.go#L30-L41
to here (the other PR was merged @stillalearner :smile: you can rebase now)

Comment (mateusoliveira43) [reply controllers/bsl.go:19]: for maintainers: should we move this code https://github.com/openshift/oadp-operator/blob/8890679c9b2c6866860badbd3c870ff3392a5ec8/controllers/validator.go#L109-L114

up in the validation chain? to do not change the order it was previously?

**Review by mateusoliveira43 — DISMISSED** [review]
Just rebase and check my last comment

Thanks for the PR @stillalearner ! :rocket:

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go:19]: @mateusoliveira43 you are right, actually this backupLocation for loop and volumeSnapshot for loop in the validator.go is also duplicate as it is being checked already in the new functions ensureDataSecretExists and the VSl part also. removing it.

Comment (stillalearner) [reply controllers/bsl.go]: done. I also think that bsl and vsl validation should be done first, as the duplicate code we are removing from there was actually doing the same previously, which is now included in ValidateBSL and ValidateVSL functions.

**Review by kaovilai — APPROVED** [review]

File: controllers/bsl.go
Comment (mpryc) [inline controllers/bsl.go:383]: Should we add on top of returning error logging error in the function like this? We could have log.error in the caller, but in my opinion it would make sense to add such error or debug to the function which actually returns the error itself, then we don't need to take care of that logging in any future caller of this func ? This would apply to other parts of this PR.

File: controllers/bsl.go
Comment (mpryc) [inline controllers/bsl.go:407]: Is there a risk of null pointer in that check? / Is there guarantee that the HasFeatureFlag always is there?

e.g.
```
if dpa.Spec != nil && dpa.Spec.Configuration != nil && !dpa.Spec.Configuration.Velero.HasFeatureFlag("no-secret") {

```

File: controllers/bsl.go
Comment (mpryc) [inline controllers/bsl.go:423]: Any type of log would be nice, so we know in the run what's wrong.

File: controllers/bsl_test.go
Comment (mpryc) [inline controllers/bsl_test.go:1744]: Would it make sense to create type of error within controllers/*.go where the strings used in errors are stored? This way we don't have to compare against string here, but the string taken from that type? In case string message changes we don't have to update it in few places.

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go:407]: should not be an issue, as dpa.Spec.Configuration and dpa.Spec.Configuration.Velero nil check are already covered in validator.go

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go:423]: I have seen this error, it is pretty straightforward when it fails, complaining that "secret <name> not found"

File: controllers/bsl_test.go
Comment (stillalearner) [reply controllers/bsl_test.go:1744]: ya this one makes sense to all the messages strings in all the test files, to be converted to Types.
We can have a separate PR for this one. I agree. Let me know if you want me to create one.

File: controllers/bsl.go
Comment (stillalearner) [reply controllers/bsl.go:383]: ya, I think this can also be part of the same PR where we can add Types for all these error strings
like,
ErrBucketConfigurationMissing = "BackupLocation must have...."

and then, adding log something like this for all functions?

`if bsl.Velero != nil && bsl.Velero.ObjectStorage != nil && bsl.Velero.ObjectStorage.Prefix == "" && dpa.BackupImages() {
		err:= fmt.Errorf("BackupLocation must have velero prefix when backupImages is not set to false")
		
		r.Log.Error(err, "Validation Failed")

		return err
}`

File: controllers/bsl_test.go
Comment (mpryc) [reply controllers/bsl_test.go:1744]: Ok so let's not this be part of current PR, but I think it makes sense to keep all the error types / strings separate. Thinking here of something like:

```
type OadpSeverity string

const (
    Error       OadpSeverity = "error"
    Warning     OadpSeverity = "warning"
    Info        OadpSeverity = "info"
    Debug       OadpSeverity = "debug"
    Deprecation OadpSeverity = "deprecation"
)

type CustomMessage struct {
    OadpMessageID   string // Unique ID to identify message
    Message         string
    Severity        Severity
    Info            map[string]interface{} // Additional information for logging or debugging
}
```

File: controllers/bsl.go
Comment (mpryc) [reply controllers/bsl.go:383]: Ok worth noting, so it's not lost, but not within current PR.

**Pull request merged**
